src/ckb-daemon/usb_linux.c

```diff
@@ -131,32 +136,37 @@ int os_usbsend(usbdevice* kb, const uchar* out_msg, int is_recv, const char* fil
 ///
 int os_usbrecv(usbdevice* kb, uchar* in_msg, const char* file, int line){
     int res;
-    // This is what CUE does, but it doesn't seem to work on linux.
-    /*if(kb->fwversion >= 0x130){
-        struct usbdevfs_bulktransfer transfer = {0};
-        transfer.ep = 0x84;
-        transfer.len = MSG_SIZE;
-        transfer.timeout = 5000;
-        transfer.data = in_msg;
-        res = ioctl(kb->handle - 1, USBDEVFS_BULK, &transfer);
-    } else {*/
+    if(kb->fwversion >= 0x120 || IS_V2_OVERRIDE(kb)){
+        // Wait for 2s, 0ns
+        struct timespec mutexwait = {2, 0};
+        if(pthread_mutex_timedlock(&kb->interruptmutex, &mutexwait) != 0){
+            pthread_mutex_unlock(&kb->interruptmutex);
+            return -1;
+        }
+        memcpy(in_msg, kb->interruptbuf, MSG_SIZE);
+        memset(kb->interruptbuf, 0, MSG_SIZE);
+        res = MSG_SIZE;
+        pthread_mutex_unlock(&kb->interruptmutex);
+    } else {
         struct usbdevfs_ctrltransfer transfer = { 0xa1, 0x01, 0x0300, kb->epcount - 1, MSG_SIZE, 5000, in_msg };
         res = ioctl(kb->handle - 1, USBDEVFS_CONTROL, &transfer);
-    //}
-    if(res <= 0){
-        ckb_err_fn("%s\n", file, line, res ? strerror(errno) : "No data read");
-        if(res == -1 && errno == ETIMEDOUT)
-            return -1;
-        else
-            return 0;
-    } else if(res != MSG_SIZE)
-        ckb_warn_fn("Read %d bytes (expected %d)\n", file, line, res, MSG_SIZE);
+        if(res <= 0){
+            ckb_err_fn("%s\n", file, line, res ? strerror(errno) : "No data read");
+            if(res == -1 && errno == ETIMEDOUT)
```

Comment:
`ckb_err_fn()` calls `fprintf()` which is allowed to set `errno`, so this branch is not taken if `fprintf` encounters an error.

Reply:
Up to you, I am just reviewing the diff.